File: firmware/src/ble_provisioning.cpp
/**
 * @file ble_provisioning.cpp
 * @brief BLE Provisioning implementation for ESP32 Pool Controller
 */

#include "ble_provisioning.h"
#include <NimBLEDevice.h>
#include <Preferences.h>
#include <WiFi.h>

// ==================== BLE UUIDs ====================
// Custom UUIDs for Pool Controller WiFi Provisioning Service
#define SERVICE_UUID        "4fafc201-1fb5-459e-8fcc-c5c9c331914b"
#define SSID_CHAR_UUID      "beb5483e-36e1-4688-b7f5-ea07361b26a8"
#define PASSWORD_CHAR_UUID  "cba1d466-344c-4be3-ab3f-189f80dd7518"
#define STATUS_CHAR_UUID    "8d8218b6-97bc-4527-a8db-13094ac06b1d"
#define NETWORKS_CHAR_UUID  "fa87c0d0-afac-11de-8a39-0800200c9a66"  // WiFi networks scan result
// Remote commands (e.g., clear WiFi). Keep in sync with dashboard JS.
#define COMMAND_CHAR_UUID   "8b9d68c4-57b8-4b02-bf19-6fd94b62f709"

// ==================== Global BLE Objects ====================
static NimBLEServer* pServer = nullptr;
static NimBLECharacteristic* pSSIDCharacteristic = nullptr;
static NimBLECharacteristic* pPasswordCharacteristic = nullptr;
static NimBLECharacteristic* pStatusCharacteristic = nullptr;
static NimBLECharacteristic* pNetworksCharacteristic = nullptr;
static NimBLECharacteristic* pCommandCharacteristic = nullptr;

// ==================== State Variables ====================
static bool bleActive = false;
static bool newCredentialsReceived = false;
static String receivedSSID = "";
static String receivedPassword = "";
static bool deviceConnected = false;
static bool clearWiFiRequested = false;

// ==================== BLE Callbacks ====================

/**
 * Server callback - handles client connect/disconnect events
 */
class ServerCallbacks : public NimBLEServerCallbacks {
  void onConnect(NimBLEServer* pServer) {
    deviceConnected = true;
    Serial.println("[BLE] Client connected");
    
    // Update status characteristic
    if (pStatusCharacteristic) {
      pStatusCharacteristic->setValue("connected");
      pStatusCharacteristic->notify();
    }
  }

  void onDisconnect(NimBLEServer* pServer) {
    deviceConnected = false;
    Serial.println("[BLE] Client disconnected");
    
    // Restart advertising so others can connect
    NimBLEDevice::startAdvertising();
    Serial.println("[BLE] Advertising restarted");
  }
};

/**
 * Characteristic callbacks - handles write events for WiFi credentials
 */
class CharacteristicCallbacks : public NimBLECharacteristicCallbacks {
  void onWrite(NimBLECharacteristic* pCharacteristic) {
    std::string uuid = pCharacteristic->getUUID().toString();
    std::string value = pCharacteristic->getValue();
    
    if (uuid == SSID_CHAR_UUID) {
      receivedSSID = String(value.c_str());
      Serial.print("[BLE] SSID received: ");
      Serial.println(receivedSSID);
      
      // Update status
      if (pStatusCharacteristic) {
        pStatusCharacteristic->setValue("ssid_received");
        pStatusCharacteristic->notify();
      }
    } 
    else if (uuid == PASSWORD_CHAR_UUID) {
      receivedPassword = String(value.c_str());
      Serial.print("[BLE] Password received (");
      Serial.print(receivedPassword.length());
      Serial.println(" chars)");
      
      // Update status
      if (pStatusCharacteristic) {
        pStatusCharacteristic->setValue("password_received");
        pStatusCharacteristic->notify();
      }
      
      // Both credentials received
      if (receivedSSID.length() > 0 && receivedPassword.length() > 0) {
        newCredentialsReceived = true;
        Serial.println("[BLE] ✓ WiFi credentials complete");
        
        if (pStatusCharacteristic) {
          pStatusCharacteristic->setValue("credentials_ready");
          pStatusCharacteristic->notify();
        }
      }
    }
    else if (uuid == NETWORKS_CHAR_UUID) {
      // Trigger WiFi scan when client writes to networks characteristic
      Serial.println("[BLE] Networks scan triggered via write");
      String json = scanWiFiNetworks();
      
      // Update the characteristic with scan results
      pCharacteristic->setValue((uint8_t*)json.c_str(), json.length());
      Serial.print("[BLE] Networks characteristic updated, length: ");
      Serial.println(json.length());
      
      // Notify client that new data is available
      pCharacteristic->notify();
    }
    else if (uuid == COMMAND_CHAR_UUID) {
      // Handle simple command verbs from dashboard
      if (value == "clear_wifi") {
        clearWiFiRequested = true;
        Serial.println("[BLE] Clear WiFi command received via BLE");

        if (pStatusCharacteristic) {
          pStatusCharacteristic->setValue("clear_wifi_requested");
          pStatusCharacteristic->notify();
        }
      }
    }
  }
  
  void onRead(NimBLECharacteristic* pCharacteristic) {
    std::string uuid = pCharacteristic->getUUID().toString();
    
    // Log when networks characteristic is read
    if (uuid == NETWORKS_CHAR_UUID) {
      Serial.println("[BLE] Networks characteristic read");
    }
  }
};

// Forward declaration for scanWiFiNetworks
String scanWiFiNetworks();

// ==================== Public Functions ====================

void initBLEProvisioning() {
  Serial.println("[BLE] Initializing BLE provisioning...");
  
  // Generate device name with MAC address suffix
  uint8_t mac[6];
  esp_read_mac(mac, ESP_MAC_WIFI_STA);
  char deviceName[32];
  // Add version suffix to break cached GATT on clients
  snprintf(deviceName, sizeof(deviceName), "Controlador Smart Pool-%02X%02X-v2", mac[4], mac[5]);
  
  Serial.print("[BLE] Device name: ");
  Serial.println(deviceName);
  
  // Initialize NimBLE
  NimBLEDevice::init(deviceName);
  
  // Create BLE Server
  pServer = NimBLEDevice::createServer();
  pServer->setCallbacks(new ServerCallbacks());
  
  // Create BLE Service
  NimBLEService* pService = pServer->createService(SERVICE_UUID);
  
  // Create SSID Characteristic (Read/Write)
  pSSIDCharacteristic = pService->createCharacteristic(
    SSID_CHAR_UUID,
    NIMBLE_PROPERTY::READ | NIMBLE_PROPERTY::WRITE
  );
  pSSIDCharacteristic->setCallbacks(new CharacteristicCallbacks());
  pSSIDCharacteristic->setValue(""); // Initial empty value
  
  // Create Password Characteristic (Write only for security)
  pPasswordCharacteristic = pService->createCharacteristic(
    PASSWORD_CHAR_UUID,
    NIMBLE_PROPERTY::WRITE
  );
  pPasswordCharacteristic->setCallbacks(new CharacteristicCallbacks());
  pPasswordCharacteristic->setValue("");
  
  // Create Status Characteristic (Read/Notify)
  pStatusCharacteristic = pService->createCharacteristic(
    STATUS_CHAR_UUID,
    NIMBLE_PROPERTY::READ | NIMBLE_PROPERTY::NOTIFY
  );
  pStatusCharacteristic->setValue("waiting");
  
  // Create Networks Characteristic (Read/Write - write triggers scan, read returns JSON)
  pNetworksCharacteristic = pService->createCharacteristic(
    NETWORKS_CHAR_UUID,
    NIMBLE_PROPERTY::READ | NIMBLE_PROPERTY::WRITE | NIMBLE_PROPERTY::NOTIFY
  );
  pNetworksCharacteristic->setCallbacks(new CharacteristicCallbacks());
  pNetworksCharacteristic->setValue("[]"); // Initial empty list

  // Create Command Characteristic (Write to request actions like clearing WiFi)
  pCommandCharacteristic = pService->createCharacteristic(
    COMMAND_CHAR_UUID,
    NIMBLE_PROPERTY::READ | NIMBLE_PROPERTY::WRITE | NIMBLE_PROPERTY::NOTIFY
  );
  pCommandCharacteristic->setCallbacks(new CharacteristicCallbacks());
  pCommandCharacteristic->setValue("");
  Serial.print("[BLE] Command characteristic UUID: ");
  Serial.println(COMMAND_CHAR_UUID);
  
  // Start the service
  pService->start();
  
  // Start advertising
  NimBLEAdvertising* pAdvertising = NimBLEDevice::getAdvertising();
  pAdvertising->addServiceUUID(SERVICE_UUID);
  pAdvertising->setScanResponse(true);
  pAdvertising->setMinPreferred(0x06);  // Apple connection parameter
  pAdvertising->setMaxPreferred(0x12);
  
  NimBLEDevice::startAdvertising();
  
  bleActive = true;
  
  Serial.println("[BLE] ✓ Provisioning service started");
  Serial.println("[BLE] Waiting for dashboard connection...");
  Serial.print("[BLE] Service UUID: ");
  Serial.println(SERVICE_UUID);
}

void stopBLEProvisioning() {
  if (!bleActive) return;
  
  Serial.println("[BLE] Stopping provisioning service...");
  
  NimBLEDevice::stopAdvertising();
  
  // deinit() automatically disconnects all clients
  NimBLEDevice::deinit(true);
  
  bleActive = false;
  deviceConnected = false;
  pServer = nullptr;
  
  Serial.println("[BLE] ✓ Provisioning stopped");
}

bool isBLEProvisioningActive() {
  return bleActive;
}

bool hasNewWiFiCredentials() {
  return newCredentialsReceived;
}

bool getBLEWiFiSSID(char* ssid) {
  if (receivedSSID.length() == 0) return false;
  
  strncpy(ssid, receivedSSID.c_str(), 32);
  ssid[32] = '\0';
  return true;
}

bool getBLEWiFiPassword(char* password) {
  if (receivedPassword.length() == 0) return false;
  
  strncpy(password, receivedPassword.c_str(), 63);
  password[63] = '\0';
  return true;
}

void clearBLECredentials() {
  newCredentialsReceived = false;
  receivedSSID = "";
  receivedPassword = "";
}

/**
 * Scan available WiFi networks and return JSON array
 * @return JSON string with network list: [{"ssid":"NETWORK1","rssi":-50,"open":false},...]
 */
String scanWiFiNetworks() {
  Serial.println("[BLE] Scanning WiFi networks...");
  
  // Ensure WiFi is in station mode for scanning (required for BLE coexistence)
  WiFi.mode(WIFI_STA);
  delay(100); // Give WiFi radio time to initialize
  
  // Perform WiFi scan
  int numNetworks = WiFi.scanNetworks();
  
  if (numNetworks == 0 || numNetworks == -1) {
    Serial.println("[BLE] No networks found or scan failed");
    return "[]";
  }
  
  Serial.print("[BLE] Found ");
  Serial.print(numNetworks);
  Serial.println(" networks");
  
  // Build JSON array
  String json = "[";
  int networkCount = 0;
  
  for (int i = 0; i < numNetworks; i++) {
    String ssid = WiFi.SSID(i);
    int rssi = WiFi.RSSI(i);
    uint8_t encryption = WiFi.encryptionType(i);
    bool open = (encryption == WIFI_AUTH_OPEN);
    
    // Skip empty SSIDs
    if (ssid.length() == 0) continue;
    
    // Build network entry
    String entry = "{\"ssid\":\"";
    entry += ssid;
    entry += "\",\"rssi\":";
    entry += String(rssi);
    entry += ",\"open\":";
    entry += (open ? "true" : "false");
    entry += "}";
    
    // Check if adding this entry would exceed safe BLE MTU (~400 bytes for reliable transmission)
    int projectedSize = json.length() + entry.length() + (networkCount > 0 ? 1 : 0) + 1; // +1 for comma, +1 for ]
    if (projectedSize > 400) {
      Serial.println("[BLE] Network list too large, stopping here");
      break;
    }
    
    if (networkCount > 0) json += ",";
    json += entry;
    networkCount++;
  }
  
  json += "]";
  
  // Clean up
  WiFi.scanDelete();
  
  Serial.print("[BLE] JSON size: ");
  Serial.print(json.length());
  Serial.println(" bytes");
  Serial.print("[BLE] JSON: ");
  Serial.println(json);
  
  return json;
}

bool isClearWiFiRequested() {
  return clearWiFiRequested;
}

void resetClearWiFiRequest() {
  clearWiFiRequested = false;
}

File: firmware/dashboard/ble-provisioning.js
  /**
   * Scan for available WiFi networks via BLE
   * Triggers scan, waits for scan to complete, and returns network list
   * @param {function} onProgress - Optional callback for scan progress
   * @returns {Promise<Array>} List of networks
   */
  async scanNetworks(onProgress) {
    if (!this.networksCharacteristic || !this.scanStatusCharacteristic) {
      throw new Error('Scan not supported on this device.');
    }

    // Clear previous results (do not persist)
    if (onProgress) onProgress('Clearing previous network list...');
    // Optionally, clear UI here

    // Write to networks characteristic to trigger scan
    if (onProgress) onProgress('Requesting WiFi scan...');
    await this.networksCharacteristic.writeValue(new Uint8Array([0]));

    // Wait for scan status to become 'done'
    let status = '';
    let maxWait = 10000; // 10s timeout
    let start = Date.now();
    while (Date.now() - start < maxWait) {
      const value = await this.scanStatusCharacteristic.readValue();
      status = new TextDecoder().decode(value);
      if (onProgress) onProgress(`Scan status: ${status}`);
      if (status === 'done') break;
      if (status === 'error') throw new Error('WiFi scan failed');
      await new Promise(resolve => setTimeout(resolve, 300));
    }
    if (status !== 'done') throw new Error('WiFi scan timed out');

    // Read network list
    const netValue = await this.networksCharacteristic.readValue();
    const json = new TextDecoder().decode(netValue);
    let networks = [];
    try {
      networks = JSON.parse(json);
    } catch (e) {
      throw new Error('Failed to parse network list');
    }
    if (onProgress) onProgress('Scan complete');
    return networks;
  },
/**
 * ESP32 Pool Controller - BLE Provisioning Module
 * Web Bluetooth API integration for WiFi credential provisioning
 * 
 * Usage:
 * 1. Include this script in your dashboard
 * 2. Call ESP32BLEProvisioning.provision() when user clicks "Add Device"
 * 3. Handle success/error callbacks
 */

const ESP32BLEProvisioning = {
  // BLE Service & Characteristic UUIDs (must match ESP32 firmware)
  SERVICE_UUID: '4fafc201-1fb5-459e-8fcc-c5c9c331914b',
  SSID_CHAR_UUID: 'beb5483e-36e1-4688-b7f5-ea07361b26a8',
  PASSWORD_CHAR_UUID: 'cba1d466-344c-4be3-ab3f-189f80dd7518',
  STATUS_CHAR_UUID: '8d8218b6-97bc-4527-a8db-13094ac06b1d',
  NETWORKS_CHAR_UUID: 'fa87c0d0-afac-11de-8a39-0800200c9a66',
  COMMAND_CHAR_UUID: '8b9d68c4-57b8-4b02-bf19-6fd94b62f709',
  SCAN_STATUS_CHAR_UUID: 'b7e1a1c2-8f8e-4e2a-9b1a-2e3b4c5d6e7f',

  // State
  device: null,
  server: null,
  service: null,
  ssidCharacteristic: null,
  passwordCharacteristic: null,
  statusCharacteristic: null,
  networksCharacteristic: null,
  commandCharacteristic: null,
  scanStatusCharacteristic: null,

  /**
   * Check if Web Bluetooth is supported
   * @returns {boolean} true if supported
   */
  isSupported() {
    if (!navigator.bluetooth) {
      console.error('[BLE] Web Bluetooth API not supported in this browser');
      return false;
    }
    return true;
  },

  /**
   * Scan for ESP32 Pool devices and connect
   * Shows browser's device picker UI
   * @returns {Promise<boolean>} true if connected successfully
   */
  async connect() {
    if (!this.isSupported()) {
      throw new Error('Web Bluetooth not supported. Use Chrome, Edge, or Opera.');
    }

    try {
      console.log('[BLE] Scanning for ESP32 Pool devices...');

      // Request device with our service UUID filter
      this.device = await navigator.bluetooth.requestDevice({
        filters: [
          { namePrefix: 'ESP32-Pool' },
          { services: [this.SERVICE_UUID] }
        ],
        optionalServices: [this.SERVICE_UUID]
      });

      console.log(`[BLE] Found device: ${this.device.name}`);

      // Connect to GATT server
      console.log('[BLE] Connecting to GATT server...');
      this.server = await this.device.gatt.connect();
      console.log('[BLE] ✓ Connected to GATT server');

      // Get service
      this.service = await this.server.getPrimaryService(this.SERVICE_UUID);
      console.log('[BLE] ✓ Got provisioning service');


      // Get characteristics
      this.ssidCharacteristic = await this.service.getCharacteristic(this.SSID_CHAR_UUID);
      this.passwordCharacteristic = await this.service.getCharacteristic(this.PASSWORD_CHAR_UUID);
      this.statusCharacteristic = await this.service.getCharacteristic(this.STATUS_CHAR_UUID);
      // Optional characteristics used by advanced flows (scan / clear commands)
      try {
        this.networksCharacteristic = await this.service.getCharacteristic(this.NETWORKS_CHAR_UUID);
      } catch (err) {
        console.warn('[BLE] Networks characteristic not found; WiFi scan over BLE will be disabled', err);
      }

      try {
        this.commandCharacteristic = await this.service.getCharacteristic(this.COMMAND_CHAR_UUID);
      } catch (err) {
        console.warn('[BLE] Command characteristic not found; remote commands like clear_wifi will be disabled', err);
      }

      try {
        this.scanStatusCharacteristic = await this.service.getCharacteristic(this.SCAN_STATUS_CHAR_UUID);
        await this.scanStatusCharacteristic.startNotifications();
        this.scanStatusCharacteristic.addEventListener('characteristicvaluechanged', (event) => {
          const value = new TextDecoder().decode(event.target.value);
          console.log(`[BLE] Scan status update: ${value}`);
          // Optionally, update UI here to show scan progress
          if (typeof window !== 'undefined' && window.updateScanStatus) {
            window.updateScanStatus(value);
          }
        });
      } catch (err) {
        console.warn('[BLE] Scan status characteristic not found; scan progress will not be shown', err);
      }

      console.log('[BLE] ✓ Got all characteristics');

      // Subscribe to status notifications
      await this.statusCharacteristic.startNotifications();
      this.statusCharacteristic.addEventListener('characteristicvaluechanged', (event) => {
        const value = new TextDecoder().decode(event.target.value);
        console.log(`[BLE] Status update: ${value}`);
      });

      return true;
    } catch (error) {
      console.error('[BLE] Connection error:', error);
      this.cleanup();
      throw error;
    }
  },

  /**
   * Send WiFi credentials to ESP32
   * @param {string} ssid - WiFi network name
   * @param {string} password - WiFi password
   * @returns {Promise<boolean>} true if sent successfully
   */
  async sendCredentials(ssid, password) {
    if (!this.server || !this.server.connected) {
      throw new Error('Not connected to device. Call connect() first.');
    }

    try {
      console.log(`[BLE] Sending SSID: ${ssid}`);
      const ssidEncoder = new TextEncoder();
      await this.ssidCharacteristic.writeValue(ssidEncoder.encode(ssid));
      console.log('[BLE] ✓ SSID sent');

      // Small delay between writes
      await new Promise(resolve => setTimeout(resolve, 100));

      console.log('[BLE] Sending password...');
      const passwordEncoder = new TextEncoder();
      await this.passwordCharacteristic.writeValue(passwordEncoder.encode(password));
      console.log('[BLE] ✓ Password sent');

      return true;
    } catch (error) {
      console.error('[BLE] Error sending credentials:', error);
      throw error;
    }
  },

  /**
   * Disconnect from device
   */
  disconnect() {
    if (this.device && this.device.gatt.connected) {
      this.device.gatt.disconnect();
      console.log('[BLE] Disconnected');
    }
    this.cleanup();
  },

  /**
   * Clean up state
   */
  cleanup() {
    this.device = null;
    this.server = null;
    this.service = null;
    this.ssidCharacteristic = null;
    this.passwordCharacteristic = null;
    this.statusCharacteristic = null;
    this.networksCharacteristic = null;
    this.commandCharacteristic = null;
  },

  /**
   * Complete provisioning flow (high-level API)
   * @param {string} ssid - WiFi network name
   * @param {string} password - WiFi password
   * @param {Object} callbacks - Optional callbacks { onProgress, onSuccess, onError }
   * @returns {Promise<void>}
   */
  async provision(ssid, password, callbacks = {}) {
    const { onProgress, onSuccess, onError } = callbacks;

    let attempt = 0;
    let delayMs = 500;
    while (attempt < 3) {
      try {
        attempt++;
        // Step 1: Connect to device
        if (onProgress) onProgress(`Scanning for ESP32 devices... (Attempt ${attempt})`);
        await this.connect();

        // Step 2: Send credentials
        if (onProgress) onProgress('Sending WiFi credentials...');
        await this.sendCredentials(ssid, password);

        // Step 3: Wait a moment for ESP32 to process
        if (onProgress) onProgress('ESP32 connecting to WiFi...');
        await new Promise(resolve => setTimeout(resolve, 3000));

        // Step 4: Disconnect
        this.disconnect();

        if (onProgress) onProgress('Provisioning complete!');
        if (onSuccess) onSuccess();

        console.log('[BLE] ✓ Provisioning completed successfully');
        return;
      } catch (error) {
        console.error(`[BLE] Provisioning attempt ${attempt} failed:`, error);
        this.disconnect();
        if (attempt >= 3) {
          if (onError) onError(error);
          throw error;
        }
        // Exponential backoff
        await new Promise(resolve => setTimeout(resolve, delayMs));
        delayMs *= 2;
      }
    }
  },

  /**
   * Request the device to forget its stored WiFi credentials
   * This writes the "clear_wifi" verb to the command characteristic.
   */
  async clearWifiCredentials() {
    if (!this.server || !this.server.connected) {
      throw new Error('Not connected to device. Call connect() first.');
    }

    if (!this.commandCharacteristic) {
      throw new Error(`Command characteristic ${this.COMMAND_CHAR_UUID} not available on this device.`);
    }

    try {
      await this.commandCharacteristic.writeValue(new TextEncoder().encode('clear_wifi'));
      console.log('[BLE] ✓ clear_wifi command sent');
    } catch (error) {
      console.error('[BLE] Failed to send clear_wifi command:', error);
      throw error;
    }
  }
};

// Export for use in modules (optional)
if (typeof module !== 'undefined' && module.exports) {
  module.exports = ESP32BLEProvisioning;
}
